source/extensions/filters/http/dynamic_cluster/dynamic_cluster.cc

```diff
@@ -0,0 +1,56 @@
+#include "extensions/filters/http/dynamic_cluster/dynamic_cluster.h"
+
+#include <arpa/inet.h>
+
+#include "common/common/assert.h"
+#include "common/common/empty_string.h"
+#include "common/common/utility.h"
+#include "common/config/cds_json.h"
+#include "common/http/headers.h"
+#include "common/http/utility.h"
+
+#include "fmt/printf.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace DynamicCluster {
+// Implements StreamDecoderFilter.
+
+Http::FilterHeadersStatus DynamicCluster::decodeHeaders(Http::HeaderMap& headers, bool) {
+
+  const Http::LowerCaseString ClusterName{"x-envoy-cluster"};
+  const Http::LowerCaseString ClusterHosts{"x-envoy-cluster-hosts"};
+  const std::string cluster_name = headers.get(ClusterName)->value().c_str();
+  const std::string cluster_hosts = headers.get(ClusterHosts)->value().c_str();
+  std::istringstream ss(cluster_hosts);
+  std::string host, port;
+  std::getline(ss, host, ':');
+  std::getline(ss, port, ':');
+
+  // TODO: Make this part of filter config - possibly multiple templates for each type http1, http2
```

Comment:
Couple of options here 
- Create Cluster object entirely based on http header information with sensible defaults to most of the fields.
- Provide filter config to select the cluster template and fill it with the fields from http headers.